### PR TITLE
Clear the snapshot fetch timer once the snapshot is fetched/failed

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Clear the snapshot fetch timer once the snapshot is fetched/failed as it can cause a memory leak.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 1]
```diff
@@ -228,7 +229,10 @@ async function fetchLatestSnapshotCore(
                     storageToken,
                     snapshotOptions,
                     controller,
-                );
+                ).finally(() => {
+                    // Clear the fetchTimeout once the response is fetched.
+                    clearTimeout(fetchTimeout);
```
Comment (marco-ms) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:234]: Should we set fetchTimeout to undefined?

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 2]
```diff
@@ -228,7 +229,10 @@ async function fetchLatestSnapshotCore(
                     storageToken,
                     snapshotOptions,
                     controller,
-                );
+                ).finally(() => {
```
Comment (marco-ms) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:232]: Does this cover the case when this async function is never called and the ODSP Driver is destroyed before that?

Comment (marco-ms) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:232]: I don't know, can we? It looks like the timer is set synchronously while its clearance asynchronously, so there is some time in between that could happen?

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:232]: The downloadSnapshot is called immediately. So, once it is called, then whether it fails/passes the timeout will be cleared. In cases where container is destroyed abruptly do the apps do that gracefully like calling close on the container? Like what I am asking is how does the driver/container knows if its destroyed or not?

Comment (Zlatkovsky) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts:232]: If an iframe were to get disposed, the `container.close()` method would get called.
For any hosts that don't use iframes, I think we can still assume that they'll call `container.close()`.

So I think Jatin's code looks good to me. It's hard to imagine a scenario where the `finally` codeblock wouldn't be getting executed...

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 3]
```diff
@@ -197,9 +197,10 @@ async function fetchLatestSnapshotCore(
...
 
         let controller: AbortController | undefined;
+        let fetchTimeout: ReturnType<typeof setTimeout> | undefined;
         if (snapshotOptions?.timeout !== undefined) {
             controller = new AbortController();
-            setTimeout(
+            fetchTimeout = setTimeout(
                 () => controller!.abort(),
                 snapshotOptions.timeout,
             );
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:206]: I'd move this block down closer to PerformanceEvent.timedExecAsync() call, to reduce amount of code in between (any exception in between will result in timer hanging around and not being cancelled)

**Review by vladsud — APPROVED** [review]

**Pull request merged**